src/ray/object_manager/common.cc

```diff
@@ -0,0 +1,160 @@
+#include "ray/object_manager/common.h"
+
+namespace ray {
+
+void PrintPlasmaObjectHeader(const PlasmaObjectHeader *header) {
+  RAY_LOG(DEBUG) << "PlasmaObjectHeader: \n"
+                 << "version: " << header->version << "\n"
+                 << "num_readers: " << header->num_readers << "\n"
+                 << "num_read_acquires_remaining: " << header->num_read_acquires_remaining
+                 << "\n"
+                 << "num_read_releases_remaining: " << header->num_read_releases_remaining
+                 << "\n"
+                 << "data_size: " << header->data_size << "\n"
+                 << "metadata_size: " << header->metadata_size << "\n";
+}
+
+void PlasmaObjectHeader::Init() {
+  // wr_mut is shared between writer and readers.
+  pthread_mutexattr_t mutex_attr;
+  pthread_mutexattr_init(&mutex_attr);
+  pthread_mutexattr_setpshared(&mutex_attr, PTHREAD_PROCESS_SHARED);
+  pthread_mutexattr_settype(&mutex_attr, PTHREAD_MUTEX_ERRORCHECK);
+  pthread_mutex_init(&wr_mut, &mutex_attr);
+
+  sem_init(&rw_semaphore, PTHREAD_PROCESS_SHARED, 1);
+
+  // Condition is shared between writer and readers.
+  pthread_condattr_t cond_attr;
+  pthread_condattr_init(&cond_attr);
+  pthread_condattr_setpshared(&cond_attr, PTHREAD_PROCESS_SHARED);
+  pthread_cond_init(&cond, &cond_attr);
+}
+
+void PlasmaObjectHeader::Destroy() {
+  RAY_CHECK(pthread_mutex_destroy(&wr_mut) == 0);
+  RAY_CHECK(pthread_cond_destroy(&cond) == 0);
+  RAY_CHECK(sem_destroy(&rw_semaphore) == 0);
+}
+
+// Get the data size of the plasma object.
+// This has to be called only when reader lock is acquired
+// via ReadAcquire.
+uint64_t PlasmaObjectHeader::GetDataSize() const {
+  RAY_CHECK_NE(num_read_releases_remaining, 0)
+      << "ReadAcquire has to be called before calling this method.";
+  return data_size;
+}
+
+void PlasmaObjectHeader::WriteAcquire(int64_t write_version,
```

Comment:
can we have unit tests for this class?

Reply:
Okay if we do this in a follow-up PR? Need to think a bit about how to structure that one, since we want to test for concurrency.

Reply:
Yea totally fine to do in a follow up!